File: src/linked-list/common/list-node.ts
// This class definition comes from the problem itself and we cannot change it, or else our submission will not be
// accepted.
export class ListNode {
  val: number;

  next: ListNode | null;

  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

export function list2array(node: ListNode | null): number[] {
  if (node === null) {
    return [];
  }

  const xs: number[] = [];
  let current: ListNode | null = node;
  while (current !== null) {
    xs.push(current.val);
    current = current.next;
  }
  return xs;
}

export function array2list(xs: number[] | null): ListNode | null {
  if (xs === null) {
    return null;
  }

  let root: ListNode | null = null;
  let current: ListNode | null = null;
  for (let i = 0; i < xs.length; i++) {
    if (current === null) {
      current = {
        val: xs[i],
        next: null
      };
      root = current;
      continue;
    }

    current.next = {
      val: xs[i],
      next: null
    };
    current = current.next;
  }

  return root;
}

File: src/leetcode/linked_list/common/list_node.py
# This class definition comes from the problem itself and we cannot change it, or else our submission will not be
# accepted.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def list2array(node: ListNode | None) -> list[int]:
    if not node:
        return []

    xs: list[int] = []
    current = node
    while current:
        xs.append(current.val)
        current = current.next

    return xs


def array2list(xs: list[int]) -> ListNode | None:
    if not xs:
        return None

    root: ListNode | None = None
    current: ListNode | None = None
    for x in xs:
        if not current:
            current = ListNode(x)
            root = current
            continue
        else:
            current.next = ListNode(x)
            current = current.next

    return root
